### ISSUE TITLE
How to use pd or adaptive algorithm in my own quadcopter?

### ISSUE THREAD

**Issue by shijy16** [bjkomer/masters-thesis#1]
Hello,I am now working at a quadcopter with a hand. In simple demo controller code given by vrep,it uses 9 as gravity compensation. But after I modified gravity_compensation in crontoller.py, it still can't run stably. Also, None of the adaptive ways worked.
 What's needed to be modified apply this controller in my quadcopter?
Thanks.

**Comment by bjkomer** [comment]
This will depend on the setup of your quadcopter. This controller was built for the quadcopter that came with V-REP version 3.2.1 and controls the four rotor speeds of that quadcopter. The model in V-REP comes with its own controller code in Lua, so you'll have to delete that so it doesn't interfere, and modify the script to be able to talk to python. Take a look at the scene file in this repo to see what is needed. If you're using a different quadcopter you'll have to make sure that the outputs are being sent to the correct rotors. You'll likely have to tune the PD gains to your quadcopter as well if you want good performance.

The adaptive component is built on top of a PD controller. By default it starts from scratch with nothing learned, so it may take a minute sitting on the ground at the start before it learns how to deal with gravity and take off. If the hand is too heavy for the strength of your rotors, that will also cause issues.

Best of luck with your project!

**Comment by shijy16** [comment]
I've tried to modify the lua scripts. I guess the problem is that the hand is unbalanced. Because after my teammate add an unbalanced landing gear on the copter. The copter shaked badly if we controlled it with the vrep controller. And when I tried to use your controller, it runs perfectly except that the copter is always  a little away from the target position (which is easy to handle).
Anyway, though I didn't find out what the bug is, my problem solved.
Thanks for your help and thesis. Good luck! :)
